### PR TITLE
feat: identify model optimizable for recursive ttu

### PR DESCRIPTION
**Pull request by adriantam**
## Description
Identify authorization models where recursive TTU can be optimized.

## References
Part of https://github.com/openfga/openfga/issues/1945

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -25,7 +25,7 @@ require (
 	github.com/natefinch/wrap v0.2.0
 	github.com/oklog/ulid/v2 v2.1.0
 	github.com/openfga/api/proto v0.0.0-20240906203051-102620ef2a66
-	github.com/openfga/language/pkg/go v0.2.0-beta.2.0.20240920201511-a36f3c71a0fe
+	github.com/openfga/language/pkg/go v0.2.0-beta.2.0.20240920211227-a5276040c542
```
Comment (adriantam) [inline go.mod:28]: Will need to update once https://github.com/openfga/language/pull/355 is merged

File: pkg/typesystem/typesystem.go [hunk 2]
```diff
@@ -414,38 +421,91 @@ func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *
 		}
 		switch curNeighbourAuthorizationModelNode.NodeType() {
 		case graph.SpecificType:
-			// TODO wildcard?? should we add a new NodeType to handle this more explicitly?'
 			curNeighbourAuthorizationModelNodeLabel := curNeighbourAuthorizationModelNode.Label()
-			if curNeighbourAuthorizationModelNodeLabel == userType {
+			if curNeighbourAuthorizationModelNodeLabel == mustHaveTerminalType {
 				hasCorrectTerminalType = true
 			} else if tuple.IsTypedWildcard(curNeighbourAuthorizationModelNodeLabel) {
```
Comment (adriantam) [inline pkg/typesystem/typesystem.go:427]: TBD for https://github.com/openfga/language/pull/356

File: pkg/typesystem/typesystem.go [hunk 3]
```diff
@@ -414,38 +421,92 @@ func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *
...
 		return false
 	}
-	return t.recursiveUsersetNodeCanFastpath(curAuthorizationModelNode, userType)
+	return t.verifyNodeEdgesOptimizable(curAuthorizationModelNode,
+		expectedEdgeAndNodeType{expectedNodeForObjectRel: curAuthorizationModelNode, expectedEdgeType: graph.DirectEdge},
+		userType)
+}
+
+// recursiveTTUNodeCanFastpath is a helper function to determine whether the node (object#relation) has
+// - there is only one edge to curAuthorizationModelNode (which is the union node)
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:475]:
```suggestion
// - there is only one incoming edge to curAuthorizationModelNode and it is the union node
```

File: pkg/typesystem/typesystem.go [hunk 4]
```diff
@@ -414,38 +421,92 @@ func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *
...
+	}
+
+	if len(childrenNodes) != 1 {
+		return false
+	}
+
+	unionNode, ok := childrenNodes[0].(*graph.AuthorizationModelNode)
+	if !ok || unionNode.Label() != "union" {
+		return false
+	}
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:493]:
```suggestion
	neighborNodesIter := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
	if neighborNodesIter.Len() != 1 {
		return false
	}
	
	neighborNodesIter.Next()

	unionNode, ok := neighborNodesIter.Node().(*graph.AuthorizationModelNode)
	if !ok || unionNode.Label() != "union" {
		return false
	}
```

**Review by miparnisari — DISMISSED** [review]

File: pkg/typesystem/typesystem.go [hunk 5]
```diff
@@ -414,38 +420,89 @@ func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *
...
+func (t *TypeSystem) recursiveTTUNodeCanFastpath(curAuthorizationModelNode *graph.AuthorizationModelNode, userType string) bool {
+	neighborNodesIter := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
+	if neighborNodesIter.Len() != 1 {
+		return false
+	}
+
+	neighborNodesIter.Next()
+
+	unionNode, ok := neighborNodesIter.Node().(*graph.AuthorizationModelNode)
+	if !ok || unionNode.Label() != "union" {
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:487]:
```suggestion
	if !ok || unionNode.Label() != UnionOperator {
```

**Review by miparnisari — APPROVED** [review]

**Review by miparnisari — APPROVED** [review]

**Pull request merged**